lib/internal/modules/esm/resolve.js

```diff
@@ -924,29 +925,36 @@ function resolveAsCommonJS(specifier, parentURL) {
     // If it is a relative specifier return the relative path
     // to the parent
     if (isRelativeSpecifier(specifier)) {
-      found = relative(parent, found);
-      // Add '.separator if the path does not start with '..separator'
+      const foundURL = pathToFileURL(found).pathname;
+      found = relativePosixPath(
+        StringPrototypeSlice(parentURL, 'file://'.length, StringPrototypeLastIndexOf(parentURL, '/')),
+        foundURL);
+
+      // Add './' if the path does not start with '../'
       // This should be a safe assumption because when loading
       // esm modules there should be always a file specified so
       // there should not be a specifier like '..' or '.'
-      if (!StringPrototypeStartsWith(found, `..${sep}`)) {
-        found = `.${sep}${found}`;
+      if (!StringPrototypeStartsWith(found, '../')) {
+        found = `./${found}`;
       }
     } else if (isBareSpecifier(specifier)) {
       // If it is a bare specifier return the relative path within the
       // module
-      const pkg = StringPrototypeSplit(specifier, '/')[0];
-      const index = StringPrototypeIndexOf(found, pkg);
+      const pkg = StringPrototypeSlice(specifier, 0, StringPrototypeIndexOf(specifier, '/'));
+      const index = StringPrototypeIndexOf(found, `node_modules${sep}${pkg}${sep}`);
       if (index !== -1) {
-        found = StringPrototypeSlice(found, index);
+        found = ArrayPrototypeJoin(
+          ArrayPrototypeMap(
+            StringPrototypeSplit(StringPrototypeSlice(found, index + `node_modules${sep}`.length), sep),
+            // Escape URL-special characters to avoid generating a incorrect suggestion
+            encodeURIComponent,
+          ),
+          '/',
```

Comment:
Shouldn't this be `sep`?

```suggestion
          sep,
```

Reply:
No, URLs uses `/` as separator, regardless of the value of `path.sep`.